// @react-spectrum/theme-light/dist/import.mjs
import "./spectrum-darkest.0d1ebb78.css";
import $fbgke$vars_spectrumdarkest_cssmodulejs from "./vars_spectrum-darkest_css.mjs";
import "./spectrum-global.8695f78d.css";
import $fbgke$vars_spectrumglobal_cssmodulejs from "./vars_spectrum-global_css.mjs";
import "./spectrum-large.16835b8e.css";
import $fbgke$vars_spectrumlarge_cssmodulejs from "./vars_spectrum-large_css.mjs";
import "./spectrum-lightest.73b0e1ef.css";
import $fbgke$vars_spectrumlightest_cssmodulejs from "./vars_spectrum-lightest_css.mjs";
import "./spectrum-medium.fff885a6.css";
import $fbgke$vars_spectrummedium_cssmodulejs from "./vars_spectrum-medium_css.mjs";


function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */ /// <reference types="css-module-types" />





let $0aa443374b81c1bd$export$bca14c5b3b88a9c9 = {
    global: ($parcel$interopDefault($fbgke$vars_spectrumglobal_cssmodulejs)),
    light: (0, ($parcel$interopDefault($fbgke$vars_spectrumlightest_cssmodulejs))),
    dark: (0, ($parcel$interopDefault($fbgke$vars_spectrumdarkest_cssmodulejs))),
    medium: ($parcel$interopDefault($fbgke$vars_spectrummedium_cssmodulejs)),
    large: ($parcel$interopDefault($fbgke$vars_spectrumlarge_cssmodulejs))
};


export {$0aa443374b81c1bd$export$bca14c5b3b88a9c9 as theme};
